{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "micro-march",
   "metadata": {},
   "source": [
    "### Sample-by-sample analysis of fetal reproductive tract "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "tough-destiny",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import os\n",
    "import sys\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "\n",
    "def MovePlots(plotpattern, subplotdir):\n",
    "    os.system('mkdir -p '+str(sc.settings.figdir)+'/'+subplotdir)\n",
    "    os.system('mv '+str(sc.settings.figdir)+'/*'+plotpattern+'** '+str(sc.settings.figdir)+'/'+subplotdir)\n",
    "\n",
    "\n",
    "sc.settings.verbosity = 3  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.settings.figdir = './figures-samples/'\n",
    "sc.logging.print_versions()\n",
    "sc.settings.set_figure_params(dpi=80)  # low dpi (dots per inch) yields small inline figures\n",
    "\n",
    "sys.executable"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fatal-activity",
   "metadata": {},
   "source": [
    "### 1. Read in the files generated by Nextflow pipeline "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "european-bandwidth",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample = 'HD_F_GON12779691'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "weighted-lexington",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '/nfs/team292/vl6/FetalReproductiveTract/RNA_QC_CellBender/data/'\n",
    "figures_dir = '/nfs/team292/vl6/FetalReproductiveTract/RNA_QC_CellBender/figures/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "controversial-shareware",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(data_dir + sample + '.h5ad')\n",
    "if 'broad_annotations' in adata.obs.columns:\n",
    "    print('ANNOTATED')\n",
    "else: \n",
    "    print('NOT ANNOTATED')\n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc90a4ac-c69b-4bbb-9ec2-9c172c483c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Median number of reads per cell: ', np.median(adata.obs['n_counts']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47ed6784-fa66-4599-8b0d-82b6d5d2eaad",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Median number of genes per cell: ', np.median(adata.obs['n_genes']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "surprised-class",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['predicted_doublets'] = adata.obs['predicted_doublets'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faced-differential",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = ['n_genes', 'n_counts', 'percent_mito', 'percent_ribo', \n",
    "                           'doublet_scores', \n",
    "                          ], ncols = 3, color_map = 'OrRd')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "interstate-wilson",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "several-terrorist",
   "metadata": {},
   "source": [
    "### 2. Load Souporcell results (if multiplexing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bearing-harassment",
   "metadata": {},
   "outputs": [],
   "source": [
    "souporcell_res = pd.read_csv('/lustre/scratch127/cellgen/cellgeni/tickets/tic-1313/results7/HD_F_GON12893251/soc/clusters.tsv', \n",
    "                            sep = '\\t', index_col = 0)\n",
    "souporcell_res.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bridal-providence",
   "metadata": {},
   "outputs": [],
   "source": [
    "souporcell_res['assignment'].value_counts(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "funded-credit",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['barcode'] = [i.split('_')[3] for i in adata.obs_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "continued-collective",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['souporcell_classification'] = adata.obs['barcode'].map(souporcell_res['status'].to_dict())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "miniature-stevens",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['souporcell_assignment'] = adata.obs['barcode'].map(souporcell_res['assignment'].to_dict())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "meaning-economics",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['souporcell_assignment'].value_counts(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "isolated-equity",
   "metadata": {},
   "outputs": [],
   "source": [
    "cross = pd.crosstab(adata.obs['souporcell_classification'], adata.obs['souporcell_assignment'])\n",
    "cross"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "concrete-boundary",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['souporcell_assignment'] = adata.obs['souporcell_assignment'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "unlimited-dance",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['souporcell_assignment'] = np.where(adata.obs['souporcell_assignment'].isin(['0/1', '1/0', '2/0', \n",
    "                                                                                      '2/1', '0/2', '1/2']), 'Doublet', adata.obs['souporcell_assignment'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "normal-forest",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['souporcell_assignment'].value_counts(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "clear-tomato",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['souporcell_assignment'] = np.where(adata.obs['souporcell_assignment'] == 'nan', 'Unassigned', adata.obs['souporcell_assignment'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "damaged-webster",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = ['n_genes', 'n_counts', 'percent_mito',\n",
    "                           'predicted_doublets', 'doublet_scores', 'souporcell_assignment',\n",
    "                          ], ncols = 3, color_map = 'OrRd', \n",
    "          save = '_'+sample+'_qc.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "undefined-execution",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove genotype doublets \n",
    "# adata = adata[[i not in ['Doublet'] for i in adata.obs['souporcell_assignment']]]\n",
    "adata = adata[[i != 1 for i in adata.obs['predicted_doublets']]]\n",
    "adata = adata[[i < 0.4 for i in adata.obs['doublet_scores']]]\n",
    "adata.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dominant-exhaust",
   "metadata": {},
   "source": [
    "### 3. Plot markers of major cell types "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "laughing-peoples",
   "metadata": {},
   "outputs": [],
   "source": [
    "utils_dir = '/nfs/team292/vl6/RepTract/'\n",
    "sys.path.append(utils_dir)\n",
    "\n",
    "import reptract_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ecdd551-791b-466b-90e5-853d53f5553e",
   "metadata": {},
   "outputs": [],
   "source": [
    "reptract_genes.markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "civic-opera",
   "metadata": {},
   "outputs": [],
   "source": [
    "for ct in reptract_genes.markers: \n",
    "    print(ct)\n",
    "    markers = [i for i in reptract_genes.markers[ct] if i in adata.var_names.to_list()]\n",
    "    if len(markers)> 0:\n",
    "        sc.pl.umap(adata, color = markers, ncols = 3,\n",
    "                   color_map = 'OrRd', use_raw = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0bb4220-aa14-416e-84f3-eb5f062a0f92",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = [i for i in mesenchymal_markers if i in adata.var_names.to_list()], ncols = 3,\n",
    "                   color_map = 'OrRd', use_raw = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "protecting-signature",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Y chromosome markers\")\n",
    "markers = [i for i in reptract_genes.markers['Male_specific'] if i in adata.var_names.to_list()]\n",
    "print(len(markers))\n",
    "if len(markers) > 0:\n",
    "    sc.pl.umap(adata, color = markers, ncols = 3,\n",
    "                   color_map = 'OrRd', use_raw = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "annoying-netscape",
   "metadata": {},
   "source": [
    "#### 4. Cluster manifold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "676bac68-b6d3-463d-880f-0eb87e661e72",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(scanpy=True, dpi=80, dpi_save=150, frameon=True, \n",
    "                     vector_friendly=True, fontsize=14, figsize=(7,7), color_map=None, \n",
    "                     format='pdf', facecolor=None, transparent=False, ipython_format='png2x')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "executed-baker",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.leiden(adata, resolution = 1, random_state = 1996)\n",
    "sc.pl.umap(adata, color = 'leiden', legend_loc = 'on data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "declared-might",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(scanpy=True, dpi=80, dpi_save=150, frameon=True, \n",
    "                     vector_friendly=True, fontsize=14, figsize=(8,6), color_map=None, \n",
    "                     format='pdf', facecolor=None, transparent=False, ipython_format='png2x')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "played-leadership",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.violin(adata, keys = ['percent_mito'], groupby = 'leiden', rotation = 90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "tough-fifteen",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.violin(adata, keys = ['n_genes'], groupby = 'leiden', rotation = 90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "electrical-vertical",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.violin(adata, keys = ['doublet_scores'], groupby = 'leiden', rotation = 90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2993c6cb-6ba1-44a7-bd4b-ab481cb6a730",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = 'leiden', legend_loc = 'on data', groups = [ '25', '26', '27'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "returning-cleanup",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Remove contaminants from the gonads and doublets\n",
    "adata = adata[[i not in ['25', '26', '27'] for i in adata.obs['leiden']]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "informative-twenty",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d28aa603-6ede-4dc8-a02b-64c67697954f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(scanpy=True, dpi=80, dpi_save=150, frameon=True, \n",
    "                     vector_friendly=True, fontsize=14, figsize=(7,7), color_map=None, \n",
    "                     format='pdf', facecolor=None, transparent=False, ipython_format='png2x')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "manufactured-kuwait",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = 'leiden', legend_loc = 'on data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "981fcff5-0dbc-4cbd-aa59-ad1a8041f4f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.leiden(adata, resolution = 1)\n",
    "sc.pl.umap(adata, color = 'leiden', legend_loc = 'on data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07448b9e-6748-4952-a360-3ebcc32eeab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.leiden(adata, resolution = 0.1, restrict_to = ('leiden', ['18']))\n",
    "sc.pl.umap(adata, color = 'leiden_R')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "annoying-packaging",
   "metadata": {},
   "source": [
    "#### 5. Find marker genes per cluster with TF-IDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adult-intake",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "bdata = anndata.AnnData(X = adata.raw.X, var = adata.raw.var, obs = adata.obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "lined-inspector",
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2.rinterface_lib.callbacks\n",
    "import logging\n",
    "# Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "import anndata2ri\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "great-breed",
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in bdata.obs.columns: \n",
    "    if col != 'leiden_R':\n",
    "        del bdata.obs[col]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d036c6a2-e514-4fb8-be7d-66015360925d",
   "metadata": {},
   "outputs": [],
   "source": [
    "bdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "executed-catering",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i bdata\n",
    "bdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "catholic-improvement",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o mrks\n",
    "\n",
    "library(SoupX)\n",
    "counts <- assay(bdata, \"X\")\n",
    "colnames(counts) <- colnames(bdata)\n",
    "rownames(counts) <- rownames(bdata)\n",
    "mrks = quickMarkers(counts, colData(bdata)$leiden_R, N = 50) # where clusters is the cell type assignment and 100 means I want the top 100 genes per cluster that pass the hypergeometric test}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "creative-immigration",
   "metadata": {},
   "outputs": [],
   "source": [
    "mrks.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "disturbed-basin",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.unique(mrks['cluster'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abroad-laugh",
   "metadata": {},
   "outputs": [],
   "source": [
    "useful = ['PITX1', 'THBS2', 'SPON2', 'PRDM8', 'PAX8', 'GATA4', 'GATA2', 'LHX9',\n",
    "          'HTR2B', 'EMX2', 'SHH', 'DLX2', 'DLX5', 'DNAH12', 'C7', 'HOXB13', 'SNX31']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acknowledged-bread",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = 'leiden', legend_loc = 'on data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "retired-movement",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster = mrks[mrks['cluster'] == '15']['gene'][0:15]\n",
    "sc.pl.umap(adata, color = [i for i in cluster if i in adata.var_names.to_list()],\n",
    "           ncols = 4, color_map = 'OrRd', use_raw = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "recent-continuity",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_markers = ['SALL1', 'MYH11', 'TP63', 'DLX5', 'SOX9', 'SHOX', 'DNAH12', 'PDZK1IP1', 'SP7', 'WIF1', 'NOTUM', 'NKD1', 'PSCA', 'KRTDAP', 'MFAP5', 'SFRP5', 'TCF21', 'FOXF1', 'ACTC1', 'OLFM3', 'PCOLCE2', 'SIX2', 'FOXD1', 'SIX1', 'NPHS1', 'NPHS2', 'AXIN2', 'EMX2', 'MYH11', 'GATA4', 'WIF1', 'SRD5A2', 'RGS5', 'BCAN', 'SP7', 'HOXC10', 'SERPINB7', 'CDH7', 'ASPN',\n",
    "           'GATA2', 'LHX9', 'NR5A1', 'SOX9', 'WNT6', 'SHISA3', 'PLAC1', 'ADH1B', 'HTR2B', 'HOXC5', \n",
    "                           'COMP', 'CRLF1', 'CALY', 'ITGA4', 'SEMA3A', 'SFRP5', 'PRDM8', 'TMTC1', \n",
    "                           'TBX5', 'TBX4', 'ITGBL1', 'LGR5', 'TSPAN8', \n",
    "                           'EMX2','UPK3B', 'TBX18', 'MSC', 'ACTA2', 'RGS5', 'PITX2', 'PTGER3', 'RAI2',\n",
    "                          'DES', 'PDGFRA', 'DCN', 'HOXB13', 'SNX31', 'TP63', 'FOXA1', 'DNAH12', 'PITX2', \n",
    "          'WIF1', 'SERPINB7', 'RORB', 'HOXA10', 'HOXA11', 'CTCN1', 'AMHR2', 'MMP28', 'ETV4', 'GRIA4', 'TMEM163', 'ZAP70', 'RUNX2', \n",
    "          'PAPP2', 'BRINP1', 'PTGES', 'ANO4', 'AR', 'HOXA13','ISL1','HOXD13', 'DLX5', 'DLX6', 'DNAH12', \n",
    "          'HSD17B2', 'HSD17B3', 'HSD17B7', 'KRT17', 'KRT14', 'KRT13', 'KRT5', 'HOXA9', 'HOXD9', 'HOXA10', 'HOXD10', \n",
    "          'HOXA11', 'HOXD11', 'HOXA7', 'HOXC6', 'HOXC8', 'AR', 'KLK3', 'SHBG', 'TMPRSS2', 'FKBP5', 'NKX3-1', 'NCOA3', 'HHIP', 'RDH10', 'THBD', 'ALX4', 'SHOX2', 'PRDM8', 'IRX1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "worst-offset",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = [i for i in random_markers if i in adata.var_names], ncols = 4, color_map = 'OrRd', use_raw = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef57e5a7-a5a7-48cd-8e84-835ac94fbdfd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "floral-bangkok",
   "metadata": {},
   "source": [
    "#### 6. Annotate clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "indoor-florence",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = 'leiden', legend_loc = 'on data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dffd884-72d2-4542-bf75-5b3cd59bb24d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = 'leiden_R', legend_loc = 'on data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "lasting-plumbing",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.shape "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "objective-burlington",
   "metadata": {},
   "outputs": [],
   "source": [
    "annots = {'0' : 'Vaginal_SmoothMuscle', '1' : 'Uterus_Ligament', '2' : 'Uterus_Ligament', \n",
    "          '3' : 'Perivascular',  \n",
    "          '4' : 'Endothelial', '5' : 'Pre-Perivascular', '6' : 'Uterus_Ligament', '7' : 'Vaginal_SmoothMuscle', \n",
    "          '8' : 'Perivascular',\n",
    "          '9' : 'Uterus_Ligament',\n",
    "          '10' : 'Uterus_SmoothMuscle', '11' : 'SchwannCell',\n",
    "          '12,0' : 'CorpusSpongiosum',  '12,1' : 'Vaginal_Mesenchyme_SFRP5', '12,2' : 'CorpusSpongiosum', '13' : 'Uterus_Mesenchyme', \n",
    "          '14' : 'Perivascular', '15,0' : 'Genital_Epidermis', '15,1' : 'KRTDAP_Epithelium', \n",
    "          '16' : 'Uterus_SmoothMuscle',\n",
    "          '17' : 'CoelomicEpithelium',\n",
    "          '18,0' : 'VaginalPlate_Epithelium', '18,1' : 'Urothelium_UPKs', \n",
    "          '18,2' : 'MüllerianVagina_Epithelium', \n",
    "          '19' : 'Endothelial',\n",
    "         '20' : 'Vaginal_SmoothMuscle', \n",
    "          '21' : 'Immune', '22' : 'Immune', \n",
    "          '23' : 'Endothelial_Lymphatic', '24' : 'Neural',  '25' : 'Coelomic_Epithelium', '26' : 'Endothelial_Lymphatic', \n",
    "          '27' : 'SchwannCell', '28' : 'Endothelial', '29' : 'Immune', '30' : 'CorpusSpongiosum',\n",
    "          '31' : 'Immune', '32' : 'Coelomic_Epithelium', '33' : 'WolffianDuct_Mesenchyme', '34' : 'Kidney_Podocyte', \n",
    "          '35' : 'VasDeferens_Epithelium', '36' : 'SchwannCell', '37' : 'MüllerianDuct_Epithelium', '42' : 'Kidney_UretericBud', \n",
    "          '39' : 'Immune', '41' : 'Endothelial', '43' : 'Endothelial', '40' : 'Kidney_SShapedBody',\n",
    "         } \n",
    "adata.obs['broad_annotations'] = adata.obs.leiden_R.map(annots)\n",
    "sc.pl.umap(adata, color = ['broad_annotations'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "macro-amateur",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = ['n_genes', 'percent_mito', 'doublet_scores', 'broad_annotations', \n",
    "                           'leiden', \n",
    "                           'souporcell_assignment'\n",
    "                          ], \n",
    "           ncols = 2, color_map = 'OrRd')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "whole-remove",
   "metadata": {},
   "source": [
    "### 5. Save annotated h5ad object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "negative-platinum",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = 'souporcell_assignment')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3859399-b512-46b3-9ad4-e7ad72289a34",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[[i != 'Unassigned' for i in adata.obs['souporcell_assignment']]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "postal-shoot",
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_dict = {'0' : 'Hrv152', '1' : 'Hrv139'}\n",
    "adata.obs['donor'] = adata.obs['souporcell_assignment'].map(donor_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cleared-seminar",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = 'donor')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "linear-booking",
   "metadata": {},
   "outputs": [],
   "source": [
    "Hrv152 = adata[[i == 'Hrv152' for i in adata.obs['donor']]]\n",
    "Hrv152.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "boxed-serum",
   "metadata": {},
   "outputs": [],
   "source": [
    "Hrv139 = adata[[i == 'Hrv139' for i in adata.obs['donor']]]\n",
    "Hrv139.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "municipal-furniture",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '/nfs/team292/vl6/FetalReproductiveTract/RNA_QC_CellBender/data/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "charming-meter",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "english-landing",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(data_dir + sample + '.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ranking-secondary",
   "metadata": {},
   "outputs": [],
   "source": [
    "Hrv139.write(data_dir + sample + '_Hrv139' + '.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "certain-antarctica",
   "metadata": {},
   "outputs": [],
   "source": [
    "Hrv152.write(data_dir + sample + '_Hrv152' + '.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "settled-collar",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "atac_env",
   "language": "python",
   "name": "atac_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
